apps/passport-client/components/screens/FrogScreens/FrogManagerScreen.tsx

```diff
@@ -0,0 +1,244 @@
+  useSyncE2EEStorage();
+  const syncSettled = useIsSyncSettled();
+
+  const [newFrogs, setNewFrogs] = useState<FrogCryptoFrogData[]>([]);
+  const [newFrogsError, setNewFrogsError] = useState<string>();
+  const handleFileChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    try {
+      const parsedData = frogParser(event.target.value);
+      setNewFrogs(parsedData);
+    } catch (error) {
+      setNewFrogsError(error.message);
+    }
+  };
+
+  const { result: frogs, isLoading, refetch, error } = useFrogs();
+
+  useEffect(() => {
+    if (error?.includes("not authorized")) {
+      window.location.replace("/");
+    }
+  }, [error]);
+
+  if (!syncSettled) {
+    return <SyncingPCDs />;
+  }
+
+  return (
+    <AppContainer bg="gray">
+      <Container>
+        <textarea onChange={handleFileChange} />
+        {newFrogsError && (
+          <ErrorMessage>Error parsing CSV: {newFrogsError}</ErrorMessage>
+        )}
+
+        {newFrogs.length > 0 && (
+          <>
+            <h2>New Frogs</h2>
+            <DataTable data={newFrogs} />
+            <button onClick={() => refetch(newFrogs)}>Add Frogs</button>
+          </>
+        )}
+
+        <h2>Frogs</h2>
+        {isLoading && <p>Loading...</p>}
+        {error && <ErrorMessage>Error fetching frogs: {error}</ErrorMessage>}
+        {!error && frogs.length === 0 && <p>No frogs</p>}
+        <DataTable data={frogs} />
+      </Container>
+    </AppContainer>
+  );
+}
+
+function useFrogs(): {
+  result: FrogCryptoFrogData[];
+  isLoading: boolean;
+  error: string | undefined;
+  refetch: (frogs: FrogCryptoFrogData[]) => void;
+} {
+  const [result, setResult] = useState<FrogCryptoFrogData[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<string>();
+
+  const identity = useIdentity();
+  const pcds = usePCDCollection();
+  const credentialCache = useCredentialCache();
+
+  const credentialManager = useMemo(
+    () => new CredentialManager(identity, pcds, credentialCache),
+    [credentialCache, identity, pcds]
+  );
+
+  const [pcd, setPcd] = useState<SerializedPCD>();
+  const [inputFrogs, setInputFrogs] = useState<FrogCryptoFrogData[]>([]);
+  useEffect(() => {
+    const fetchPcd = async () => {
+      const pcd = await credentialManager.requestCredential({
+        signatureType: "sempahore-signature-pcd"
+      });
+      setPcd(pcd);
+    };
+    fetchPcd();
+  }, [credentialManager]);
+
+  const req = useMemo(
+    () => ({
+      pcd,
+      frogs: inputFrogs
+    }),
+    [inputFrogs, pcd]
+  );
+
+  useEffect(() => {
+    const abortController = new AbortController();
+
+    const fetchFrogs = async () => {
+      if (!req.pcd) {
+        setError("Waiting for PCD to be ready");
+        return;
+      }
+
+      try {
+        setError(undefined);
+        setIsLoading(true);
+        const result = await requestFrogCryptoManageFrogs(
+          appConfig.zupassServer,
+          req
+        );
```

Comment:
this is kinda sketching me out. this may not be what's happening, but given the way you structured the code here, here's something I am worried about happening:

- page is loaded
- `inputFrogs` is `[]` to begin with
- `useEffect` on line 113 is invoked
- `requestFrogCryptoManageFrogs` is called with `frogs` being `[]`
- i.e. simply loading this page has a chance of wiping the frogs?

edit: having looked at the server-side end of this code, looks like the `manageFrogs` endpoint is an upsert. idk if it should be an upsert or replace, slightly prefer replace maybe ... ? in the case that we make it a replace, this code is still sketchy

Reply:
Typed this more explicitly to handle three cases 1/ initial load 2/ update/replace/insert 3/ delete